### PR TITLE
Feature/mc 463 add cachedschematicgenerator

### PR DESCRIPTION
**Pull request by mc-dorzo**

### CODE REVIEW

File: src/parlant/adapters/vector_db/transient.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        self._embedder_factory = embedder_factory
+        self._embedder_type = embedder_type
+
+        self._multi_tenant: nano_vectordb.MultiTenantNanoVDB
+        self._collection_name_to_tenant_id: dict[str, str] = {}
+        self._collections: dict[str, _TransientCollection[BaseDocument]] = {}
+
+    async def __aenter__(self) -> Self:
+        self._embedder = self._embedder_factory.create_embedder(self._embedder_type)
+        self._multi_tenant = nano_vectordb.MultiTenantNanoVDB(self._embedder.dimensions)
```
Comment (kichanyurd) [inline src/parlant/adapters/vector_db/transient.py:56]: Why do we need the multi tenant version here?

File: src/parlant/adapters/vector_db/transient.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    async def __aexit__(
+        self,
+        exc_type: Optional[type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[object],
+    ) -> None:
+        pass
+
+    @override
+    async def create_collection(
```
Comment (kichanyurd) [inline src/parlant/adapters/vector_db/transient.py:68]: You're returning a private class. It's a design bug to expose a private class. Either return `VectorCollection` or make `TransientVectorCollection` public (no underscore prefix).

File: tests/test_utilities.py [hunk 3]
```diff
@@ -35,10 +52,34 @@
...
+    Session,
+    SessionId,
+    SessionStore,
+)
 from parlant.core.tools import LocalToolService, ToolId, ToolResult
+from parlant.core.persistence.common import ObjectId
+from parlant.core.persistence.document_database import DocumentCollection
 
 T = TypeVar("T")
+GLOBAL_CACHE_FILE = Path("_schematic_cache.json")
```
Comment (kichanyurd) [inline tests/test_utilities.py:77]: Not standard form to name the file with an underscore prefix. Files aren't python attributes :)
Let's call it `schematic_generation_cache.json`

File: tests/test_utilities.py [hunk 4]
```diff
@@ -35,10 +52,34 @@
...
+CACHE_SCHEMATIC_GENERATION_RESULTS = (
+    True
+    if os.environ.get("CACHE_SCHEMATIC_GENERATION_RESULTS", "True").lower() == "true"
```
Comment (kichanyurd) [inline tests/test_utilities.py:80]: Why do we need this?

File: tests/test_utilities.py [hunk 5]
```diff
@@ -312,3 +353,150 @@ def get_when_done_or_timeout(
         sleep(1)
 
     raise TimeoutError()
+
+
+TBaseModel = TypeVar("TBaseModel", bound=DefaultBaseModel)
+
+
+class _SchematicGenerationResultDocument(TypedDict, total=False):
```
Comment (kichanyurd) [inline tests/test_utilities.py:361]: Not sure why we need the whole shabang with `DocumentDatabase` and models here. Why not just `json.dump/load` to a dict?

Comment (mc-dorzo) [reply tests/test_utilities.py:361]: Instead start to manage inside the transient/flush all over again (I don't want every call for generation to start json.dump/load, its just takes more time, and with this implementation its more clean and easier. I actually started with json.dump/load and moved to this once I understand that I dont want to load it everytime..

**Pull request merged**